code/game/objects/effects/effect_system/fluid_spread/effects_foam.dm

```diff
@@ -0,0 +1,421 @@
+// Foam
+
+/// The minimum foam range required to start diluting the reagents past the minimum dilution rate.
+#define MINIMUM_FOAM_DILUTION_RANGE 3
+/// The minumum foam-area based divisor used to decrease foam exposure volume.
+#define MINIMUM_FOAM_DILUTION DIAMOND_AREA(MINIMUM_FOAM_DILUTION_RANGE)
+///	The effective scaling of the reagents in the foam. (Total delivered at or below [MINIMUM_FOAM_DILUTION])
+#define FOAM_REAGENT_SCALE 3.2
+
+/**
+ * ## Foam
+ *
+ * Similar to smoke, but slower and mobs absorb its reagent through their exposed skin.
+ */
+/obj/effect/particle_effect/fluid/foam
+	name = "foam"
+	icon_state = "foam"
+	opacity = FALSE
+	anchored = TRUE
+	density = FALSE
+	layer = EDGED_TURF_LAYER
+	plane = GAME_PLANE_UPPER
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	animate_movement = NO_STEPS
+	/// The types of turfs that this foam cannot spread to.
+	var/static/list/blacklisted_turfs = typecacheof(list(
+		/turf/open/space/transit,
+		/turf/open/chasm,
+		/turf/open/lava,
+	))
+	/// The typepath for what this foam leaves behind when it dissipates.
+	var/atom/movable/result_type = null
+	/// Whether or not this foam can produce a remnant movable if something of the same type is already on its turf.
+	var/allow_duplicate_results = TRUE
+	/// The amount of time this foam stick around for before it dissipates.
+	var/lifetime = 8 SECONDS
+	/// Whether or not this foam should be slippery.
+	var/slippery_foam = TRUE
+
+
+/obj/effect/particle_effect/fluid/foam/Initialize(mapload)
+	. = ..()
+	create_reagents(1000)
+	START_PROCESSING(SSfastprocess, src)
```

Comment:
Put these on their own subsystem, smoke or foam dying should not cause everything fast processing to choke

Reply:
Sure. Separate for smoke and foam or both on the same system. In the latter case I may have to learn how to bucket.

Reply:
separate would be best I think. I might end up asking you to make foam more advanced, I don't like the way it handles growth that much

Reply:
Neither do I. I'm about 90% certain the foam spread algorithm is the reason the vent foam event ground the servers to dust.

Reply:
All of the foam trying to spread to everywhere around itself every tick.

Reply:
Yeah. That sounds about right. At the very least I'd like it to be better queuable, though we may need to track edges too. I'd like vent foam back.